h2/src/main/org/h2/expression/function/ToChar.java

```diff
@@ -546,6 +546,54 @@ private static String getTimeZone(Session session, Value value, boolean tzd) {
         }
     }
 
+    private static String getTimeZoneHours(Session session, Value value) {
+        if(value instanceof ValueTimestampTimeZone) {
+            return printTimeZoneHoursFromOffsetSeconds(((ValueTimestampTimeZone) value).getTimeZoneOffsetSeconds());
+        } else if( value instanceof ValueTimeTimeZone) {
+            return printTimeZoneHoursFromOffsetSeconds( ((ValueTimeTimeZone)value).getTimeZoneOffsetSeconds());
+        } else {
+            TimeZoneProvider tz = session.currentTimeZone();
+            ValueTimestamp v = (ValueTimestamp) value.convertTo(TypeInfo.TYPE_TIMESTAMP, session);
+            return printTimeZoneHoursFromOffsetSeconds(tz.getTimeZoneOffsetLocal(v.getDateValue(), v.getTimeNanos()));
+        }
```

Comment:
1. Increase visibility of `DateTimeFunctions.extractIntegerField()` and use it here (`DateTimeFunctions.extractIntegerField(session, DateTimeFunctions.TIMEZONE_HOUR, value);`.

2. Pass `StringBuilder` directly to this method and to `printTimeZoneHoursFromOffsetSeconds()`.

3. Inline `printTimeZoneHoursFromOffsetSeconds()` into this method.

4. If resulting method will be not very large, inline it into `toCharDateTime()`.